libcxx/src/debugging.cpp

```diff
@@ -0,0 +1,184 @@
+    _LIBCPP_ASSERT_INTERNAL(false, "Could not read from '/proc/{pid}/status'.");
+    return false;
+  }
+
+  if (__status.pr_flag & STRC)
+    return true;
+
+  return false;
+}
+
+#elif defined(__APPLE__) || defined(__FreeBSD__)
+
+// Returns true if the current process is being debugged (either
+// running under the debugger or has a debugger attached post facto).
+static bool __is_debugger_present() noexcept {
+  // Technical Q&A QA1361: Detecting the Debugger
+  // https://developer.apple.com/library/archive/qa/qa1361/_index.html
+
+  // Initialize mib, which tells 'sysctl' to fetch the information about the current process.
+
+  array __mib{CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()};
+
+  // Initialize the flags so that, if 'sysctl' fails for some bizarre
+  // reason, we get a predictable result.
+
+  struct kinfo_proc __info {};
+
+  // Call sysctl.
+  // https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/sysctl.3.html
+
+  size_t __info_size = sizeof(__info);
+  if (sysctl(__mib.data(), __mib.size(), &__info, &__info_size, nullptr, 0) != 0) {
+    _LIBCPP_ASSERT_INTERNAL(false, "'sysctl' runtime error");
+    return false;
+  }
+
+  // If the process is being debugged if the 'P_TRACED' flag is set.
+  // https://github.com/freebsd/freebsd-src/blob/7f3184ba797452703904d33377dada5f0f8eae96/sys/sys/proc.h#L822
+
+#  if defined(__FreeBSD__)
+  const auto __p_flag = __info.ki_flag;
+#  else // __APPLE__
+  const auto __p_flag = __info.kp_proc.p_flag;
+#  endif
+
+  return ((__p_flag & P_TRACED) != 0);
+}
+
+#elif defined(__linux__)
+
+static bool __is_debugger_present() noexcept {
+#  if defined(_LIBCPP_HAS_NO_FILESYSTEM)
+  _LIBCPP_ASSERT_INTERNAL(false,
+                          "Function is not available. Could not open '/proc/self/status' for reading, libc++ was "
+                          "compiled with _LIBCPP_HAS_NO_FILESYSTEM.");
+  return false;
+#  else
+  // https://docs.kernel.org/filesystems/proc.html
+
+  // Get the status information of a process by reading the file /proc/PID/status.
+  // The link 'self' points to the process reading the file system.
+  FILE* __proc_status_fp = fopen("/proc/self/status", "r");
+  if (__proc_status_fp == nullptr) {
+    _LIBCPP_ASSERT_INTERNAL(false, "Could not open '/proc/self/status' for reading.");
+    return false;
+  }
+
+  char* __line               = nullptr;
+  size_t __lineLen           = 0;
+  const char* __tokenStr     = "TracerPid:";
+  bool __is_debugger_present = false;
+
+  while ((getline(&__line, &__lineLen, __proc_status_fp)) != -1) {
+    // If the process is being debugged "TracerPid"'s value is non-zero.
+    char* __tokenPos = strstr(__line, __tokenStr);
+    if (__tokenPos == nullptr) {
+      continue;
+    }
+
+    __is_debugger_present = (atoi(__tokenPos + strlen(__tokenStr)) != 0);
+    break;
+  }
+
+  free(__line);
+  fclose(__proc_status_fp);
+
+  return __is_debugger_present;
```

Comment:
The paper has the non-normative note "On POSIX this can be achieved by checking for a tracer parent process, with best effort determination that such a tracer parent process is a debugger." Did you investigate how easy it is to determined the process name of the tracing program and add a filter for some names?

Reply:
`atoi(__tokenPos + strlen(__tokenStr))` gives you the tracer's PID so you can use `/proc/{pid}` again to find its details (although not 100% reliably if the executable has been removed from the filesystem, or the tracer process modified its `argv[0]`).